drivers/ethernet/eth_liteeth.c

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (c) 2019 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME eth_liteeth
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <kernel.h>
+#include <device.h>
+#include <soc.h>
+#include <stdbool.h>
+#include <net/ethernet.h>
+#include <net/net_if.h>
+#include <net/net_pkt.h>
+
+#include <misc/printk.h>
+
+/* flags */
+#define LITEETH_EV_TX		0x1
+#define LITEETH_EV_RX		0x1
+
+/* slots */
+#define LITEETH_SLOT_BASE	DT_INST_0_LITEX_ETH0_BASE_ADDRESS_0
+#define LITEETH_SLOT_RX0	((LITEETH_SLOT_BASE) + 0x0000)
+#define LITEETH_SLOT_RX1	((LITEETH_SLOT_BASE) + 0x0800)
+#define LITEETH_SLOT_TX0	((LITEETH_SLOT_BASE) + 0x1000)
+#define LITEETH_SLOT_TX1	((LITEETH_SLOT_BASE) + 0x1800)
+
+/* sram - rx */
+#define LITEETH_RX_BASE		DT_INST_0_LITEX_ETH0_BASE_ADDRESS_1
+#define LITEETH_RX_SLOT		((LITEETH_RX_BASE) + 0x00)
+#define LITEETH_RX_LENGTH	((LITEETH_RX_BASE) + 0x04)
+#define LITEETH_RX_EV_PENDING	((LITEETH_RX_BASE) + 0x28)
+#define LITEETH_RX_EV_ENABLE	((LITEETH_RX_BASE) + 0x2c)
+
+/* sram - tx */
+#define LITEETH_TX_BASE		((DT_INST_0_LITEX_ETH0_BASE_ADDRESS_1) + 0x30)
+#define LITEETH_TX_START	((LITEETH_TX_BASE) + 0x00)
+#define LITEETH_TX_READY	((LITEETH_TX_BASE) + 0x04)
+#define LITEETH_TX_SLOT		((LITEETH_TX_BASE) + 0x0c)
+#define LITEETH_TX_LENGTH	((LITEETH_TX_BASE) + 0x10)
+#define LITEETH_TX_EV_PENDING	((LITEETH_TX_BASE) + 0x1c)
+
+/* irq */
+#define LITEETH_IRQ		DT_INST_0_LITEX_ETH0_IRQ_0
+#define LITEETH_IRQ_PRIORITY	CONFIG_ETH_LITEETH_0_IRQ_PRI
+
+/* label */
+#define LITEETH_LABEL		DT_INST_0_LITEX_ETH0_LABEL
+
+struct eth_liteeth_dev_data {
+	struct net_if *iface;
+	u8_t mac_addr[6];
+};
+
+struct eth_liteeth_config {
+	void (*config_func)(void);
+};
+
+static u8_t txslot;
```

Comment:
What about putting these four slot variables into the eth_liteeth_devdata struct, then all the data related to one device instance is there. Also it makes it easier to have multiple instance of this driver in the system (although I am not sure if you could have two controllers in this system).

Reply:
Good point. I moved them.